### PR TITLE
Use new Breeze for building, pulling and verifying the images.

### LINKED ISSUE #20961: CI: Rewrite waiting for CI images in Python
There is a script that waits for CI images being available. What it does it tries to pull the images with specific COMMIT_HASH id and runs tests for those image after they are downloaded.

It runs in a loop and checks if the images are ready.

It uses common "libraries" to puill the images - it will also rebuild the images if needed, but it should be rewrittten to just do that one thing: pull the images and run checks when they are complete. 

The tricky part here is that pulling the images is done in parallell to speed it up. But it does not have to really. The images can be pulled one-by-one, there is not big speedup if they are pulled in parallell I believe (`docker pull` parallelizes the pull well enough)

### LINKED ISSUE #22825: Breeze: Provider package documentation integration with Breeze
We have "Prepare provider documentation" action in Breeze that builds the documentation via running appropriate Python script and passing appropriate parameters. 

It shoudl be added to the new Breeze.

### LINKED ISSUE #22826: Breeze: Provider package building integration with Breeze
There is a command to build provider packages with Breeze image by running a python script with parameters  in the container. It should be added to the new  Breeze.

### LINKED ISSUE #22828: Breeze: Add Generate constraints in Breeze
There is a command to upgrade constraints in Breeze. It should take parameters instead of passing variables (or actually do both) in the new Breeze..

### LINKED ISSUE #22829: Breeze: Building airflow packages with Breeze
There is a command to build provider packages with Breeze image by running a python script with parameters in the container. It should be added to the new Breeze.

### LINKED ISSUE #23076: Rewrite image verification in Breeze
Image verification is run as separate command with new virtualenv, but it should be used in CI from Breeze

### LINKED ISSUE #23077: Rewrite image waiting in Breeze
Image waiting usess GNU parallell and it's running docker pull and verification  in parallel. This should all be rewritten for Breeze.

### LINKED ISSUE #23102: Rewrite rest of test-docker-compose-quickstart in python
The docker-compose-quickstart tests should be run via Breeze.

### PR DESCRIPTION
**Pull request by potiuk**
We have the new Python-based breeze and we want to replace all
the functionality used from the old Breeze with the new one.

The most important is image management. It is used in many
places and this PR replaces all the places and removes image
builidng, pushing, pulling to use the new Breeze everywhere.

That includes:

* Building and pushing CI image on CI
* Building and pushing PROD image on CI which includes:
  * building Airflow packages
  * building Provider packages
  * waiting for images in parallel (both CI and and PROD)
  * veifying the images (both CI and PROD)

All those commands have been moved to breeze.py and that required
some modification in the Breeze parameter handling - mainly related
to adding spaces when help was displayed because long list of
packages looked very bad in help output.

It's been easier to implement it in one big PR as using
image building and pulling was deeply embedded in many scripts.

With this change all the scripts in CI use directly breeze
and combine using command line parameters with evn variables
directly in the job that execute the breeze commands which
makes it much easier to understand what is going on and
repeat it locally using Breeze.

Fixes: #22825
Fixes: #23077
Fixes: #23076
Fixes: #22829
Fixes: #22828
Fixes: #22826
Fixes: #20961
Fixes: #23102

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: dev/REFRESHING_CI_CACHE.md [hunk 1]
```diff
@@ -91,18 +91,57 @@ docker run --privileged --rm tonistiigi/binfmt --install all
...
 More information can be found [here](https://docs.docker.com/engine/reference/commandline/buildx_create/)
 
-The images can be rebuilt and refreshed after the constraints are pushed. Refreshing image for particular
-python version is a simple as running the [refresh_images.sh](refresh_images.sh) script with python version
-as parameter:
+However, emulation is very slow - more than 10x slower than hardware-backed builds.
+
+## Setting up cache refreshing with hardware ARM/AMD support
+
+If you plan to build  a number of images, probably better solution is to set up a hardware remote builder
```
Comment (eladkal) [inline dev/REFRESHING_CI_CACHE.md:98]:
```suggestion
If you plan to build a number of images, probably better solution is to set up a hardware remote builder
```

**Comment by potiuk** [comment]
Apologies for such a HUGE PR, but I could not find a better way of replacing old with  new Breeze's  build/pull/push image.

I could keep on replacing piece by piece but keeping the two in parallel in CI was really burdensome. That's why I replaced all the related build/pull/push tasks in one go. I am still testing it (I will do some tests with "main" push to my own repo).  But it should be generally working.

It removes another 1000 lines of bash code (YAY!) and paves the way to remove nearly all of them in subsequent PRs. 

Most of all It unifies the way how CI and local development works - rather than having scripts on CI and local breeze commands, everythong is done via python-based breeze commands (including running stuff in parallel on CI which does not use Gnu Parallel but Python Multiprocessing)

It's way neater, nicere and much easier to reproduce locally. Example - most complex "building PROD image" is this:

````yaml
      - name: "Build & Push PROD image ${{ matrix.python-version }}:${{ env.IMAGE_TAG_FOR_THE_BUILD }}"
        run: |
          rm -f "./dist/"*.{whl,tar.gz}
          rm -f "./docker-context-files/"*.{whl,tar.gz}
          breeze prepare-provider-packages \
              --package-list-file scripts/ci/installed_providers.txt" \
              --package-format wheel
          breeze prepare-airflow-package --package-format wheel
          mv "./dist/"*.{whl,tar.gz} "./docker-context-files"
          breeze build-prod-image --push-image --install-from-docker-context-files
        env:
          PYTHON_MAJOR_MINOR_VERSION: ${{ matrix.python-version }}
          UPGRADE_TO_NEWER_DEPENDENCIES: ${{ needs.build-info.outputs.upgradeToNewerDependencies }}
          DOCKER_CACHE: ${{ needs.build-info.outputs.cacheDirective }}
          IMAGE_TAG: ${{ env.IMAGE_TAG_FOR_THE_BUILD }}
````

Which you can very easily reproduce locally with the new breeze. It combines (for clarity of the commands) both evn-vars and flags, but all the relevant env vars used are now added directly in each step rather than in job or in workflow, which makes it much more obvious what's going on.

@mik-laj  - I also removed the "special" preparation of virtualenv for docker tests - the docker tests now can be executed via `breeze` command and they will automatically use breeze's virtualenv (which already containts pytest, pytest-xdistts and requests):

**Comment by potiuk** [comment]
This is the currernt set of commands supported now: 

image

File: tests/system/providers/google/bigquery/example_bigquery_sensors.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 from datetime import datetime
 
 from airflow import models
+from airflow.models.baseoperator import BaseOperator
```
Comment (eladkal) [inline tests/system/providers/google/bigquery/example_bigquery_sensors.py:26]: Why changing system tests is part of this PR?

**Comment by potiuk** [comment]
cc: @Bowrna @edithturn -> getting much closer to replace all of it with the new Breeze :)

**Review by eladkal — APPROVED** [review]
This is massive!
🎉

File: tests/system/providers/google/bigquery/example_bigquery_sensors.py (hunk 2 above)
Comment (potiuk) [reply tests/system/providers/google/bigquery/example_bigquery_sensors.py:26]: this came as part of the semi-rndom "mypy" failures. Mypy detected slightly different errors when there was a different set of files - so I had some 20 more mypy failures with this change. 

 I thought it has been merged separately but I will definitely check it (I will myself do a thorough review of it before I merge :).

**Comment by potiuk** [comment]
> This is massive! tada

Oh yeah. Unfortunately. Now You know why your PRs were neglected a bit recently ;). The nice thing about it that it really implements all the remaining bits of the "breeze" infrastructure (for example running stuff in parallel is now embedded in some Breeze's commands which make the maintenance stuff much nicer and easier to perform by anyone without too much of a learning curve. After we finish - pretty much everything what happens in CI will be just the right sequence of breeze commands.

**Comment by potiuk** [comment]
All right. All green. Merging.

I tested:

* PR coming from fork here
* main build here: https://github.com/potiuk/airflow/actions/runs/2210724864
* PR coming from airflow here: https://github.com/apache/airflow/pull/23150

That should cover all scenarios - there is slight chance of problems with cache refreshing but we can handle it if we see a problem.

**Pull request merged**